docs/release-notes/mongoid-8.0.txt

```diff
@@ -57,6 +57,45 @@ changed in Mongoid 8.0:
 Please refer to :ref:`configuration option <configuration-options>` for
 the description and effects of each of these options.
 
+Storing Uncastable Value
+------------------------
+
+In Mongoid 8, Mongoid standardizes the storing of "uncastable values." On
+attempting to write an uncastable value, a ``nil`` is written instead. See the
+secion on :ref:`Uncastable Values <uncastable-values>` for more details.
+
+Some ``mongoize`` methods were also changed to perform consistently with rails
+and the other mongoize methods. The following is a table of the changes in
+functionality:
+
++--------------+------------------------+------------------------+-------------------+
+| Field Type   | Situation              | Previous Functionality | New Functionality |
++==============+========================+========================+===================+
+| Boolean      | When a non-boolean     | return ``false``       | return ``nil``    |
+|              | string is assigned:    |                        |                   |
+|              | "bogus value"          |                        |                   |
++--------------+------------------------+------------------------+-------------------+
+| Symbol       | When a value that does | return ``nil``         | return ``nil``    |
+|              | not respond to         |                        |                   |
+|              | ``to_sym`` is          |                        |                   |
+|              | assigned: ``[]``       |                        |                   |
++--------------+------------------------+------------------------+-------------------+
+| Array/Set    | When a value that is   | return ``[ 1, 2, 3 ]`` | return ``nil``    |
+|              | not an array or set is | return ``[[ :x, 1 ]]`` | return ``nil``    |
+|              | assigned, but does     |                        |                   |
+|              | respond to ``to_a``:   |                        |                   |
+|              | ``1..3``, ``{ x: 1 }`` |                        |                   |
```

Comment:
If this was the legacy behavior for `Array/Set`, I really think we should keep this. It may potentially break apps otherwise.

Reply:
Johnny, you wanted Mongoid to silently discard input data. For erroring on unconvertable values we have https://jira.mongodb.org/browse/MONGOID-5382.

Range and hash are not sensible inputs for an array field.

Reply:
> Range and Hash are not sensible inputs for an array field

They may not be *common*, but they are *sensible*. Reasons to keep:
1. Range/Hash both define `.to_a`, which makes users assume that Mongoid will cast them. (Similar to `.to_i`, `.to_s`, `.to_sym`, etc.)
2. Range/Hash casting to Array is always loss-less.
3. Casting Ranges/Hashes to Array is fairly common in normal Ruby code `digits = (0..9).to_a`
4. It is supported today; for above reasons some users may rely on it. Converting to nil could cause data-loss (potentially severe issue) in existing apps, so *any removal **must** have a feature flag.*

> you wanted Mongoid to silently discard input data

We should *not* discard data if it can be casted in a lossless/unambiguous manner. Range/Hash-to-Array are lossless. Consider this example:

```ruby
class Person
  include Mongoid::Document

  field :name_ary, type: Array # [first, last]
  field :age, type: Integer
end

name_hash = { "Bob" => "Smith" } # oops!
person.name_ary = name_hash
person.save!

# Although name wasn't an Array, it's better to save the data rather than discarding it.
person.name_ary #=> ["Bob", "Smith"]

person.age = "Foobar"
person.save!

# "Foobar" is completely non-sensical as an Integer; we have to discard it.
person.age #=> nil
```

Now, please let me further clarify my position:
- Range/Hash-to-Array are "castable", so let's cast them.
- For bona-fide "uncastable" types, given a choice between (a) raising an error and (b) discarding data, we should discard data.
- However, this is a *false choice*. For "uncastable", the best thing to do would be (c) introduce a `Mongoid::UncastableValue` wrapper type. On *mongoization* (presistence), this should output a warning log line, and on demongoization (retrieval) it should return the value wrapped by `Mongoid::UncastableValue`, and allow user to call `Mongoid::UncastableValue#raw_value` to handle it.